sklearn/impute.py

```diff
@@ -138,30 +185,53 @@ def fit(self, X, y=None):
         -------
         self : SimpleImputer
         """
-        # Check parameters
-        allowed_strategies = ["mean", "median", "most_frequent"]
-        if self.strategy not in allowed_strategies:
-            raise ValueError("Can only use these strategies: {0} "
-                             " got strategy={1}".format(allowed_strategies,
-                                                        self.strategy))
+        X = self._validate_input(X)
+
+        # default fill_value is 0 for numerical input and "missing_value"
+        # otherwise
+        if self.fill_value is None:
+            if X.dtype.kind in ("i", "u", "f"):
+                fill_value = 0
+            else:
+                fill_value = "missing_value"
+        else:
+            fill_value = self.fill_value
+
+        # fill_value should be numerical in case of numerical input
+        if self.strategy == "constant":
+            if X.dtype.kind in ("i", "u", "f"):
+                if not isinstance(fill_value, numbers.Real):
+                    raise TypeError(
+                        "'fill_value'={0} is invalid. Expected a numerical"
+                        " value when imputing numerical"
+                        " data".format(fill_value))
+
+            elif X.dtype.kind == "O":
```

Comment:
Hmmm... object dtype does not necessarily mean string data. And string data does not necessarily mean object dtype (numpy.asarray with strings will not result in object data; we need to test this case). So I'm not sure this validation or its error message is quite right. I don't think we need to support numeric data in object arrays necessarily, but we do need to support string arrays if we're going to support object arrays, IMO, and we need to make sure the error messages are appropriate to the various cases.

Reply:
+1 for adding support and tests for accepting string arrays even if in most cases that are not a recommended data structure for representing categorical variables.

As for the input validation on object dtype array, we can just drop the `elif X.dtype.kind == "O"` case and not raise any `TypeError` and accept any `fill_value` when `X.dtype.kind == "O"`.

Reply:
Note that there are some complexities to deal with in that case, eg the fact that filling with new values might truncate silently:

```
In [5]: a = np.array(['a', 'b', 'a'])

In [6]: mask = a == 'b'

In [7]: a[mask] = 'missing'

In [8]: a
Out[8]: 
array(['a', 'm', 'a'],
      dtype='<U1')
```

But that's then up to the user's responsibility?

Reply:
> object dtype does not necessarily mean string data

On this part (so not about supporting string dtype), I think we briefly discussed this and our feeling was: "let's be more restrictive than theoretically needed for now, if there is request for it we can always relax later". 

For example, a situation where you can get the case of array of object dtype consisting of mixed types (and not only strings), is if you pass a subset of columns of a dataframe with both string and numerical columns to the SimpleImputer. But, normally you will want a different fill value for both types of columns, and you will need a ColumnTransformer anyhow to have a specific SimpleImputer for each type of columns.

Reply:
Can't we force `dtype = object` as soon as  input dtype is not numeric ?
Something like:
```python
if X.dtype.kind not in ("i", "f"):
    X = X.astype(object)
```
right after `check_array` ?

Reply:
Thanks @jorisvandenbossche, indeed imputing a string array is not simple to implement and probably useless in practice.

I think I prefer to raise an informative error message asking the user to provide either numerical data with integer or floating point data types or categorical represented with integer or object datatypes.

If we discover a common pipeline where the implicit conversion from string dtype to object dtype suggested above by @jeremiedbb is helpful we might want to consider it later, but to me it sounds like a [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it).